File: PA24.java
//Leetcode 740. Delete and Earn
//Time complexity : O(2p)+O(k) == O(n) // p is no of elements in original array 'nums' while k is the max element in nums
//Space Complexity : O(k) == O(n)  //k is the max element in nums
class Solution {
    public int deleteAndEarn(int[] nums) {
        int max=0;
        for(int i=0; i<nums.length; i++){
            if(max<nums[i]){
                max=nums[i];
            }
        }
        int[] dp = new int[max+1];
         for(int i=0; i<nums.length; i++){
             dp[nums[i]]= dp[nums[i]]+nums[i];
         }
        for(int j=2; j<dp.length; j++){
            dp[j]=Math.max(dp[j-1] , (dp[j-2]+dp[j]));
        
        }
        return dp[max];
    }
}

File: PA_24.py
#LC 740 - Delete and Earn
#Time Complexity - O(n+max(arr))
#Space Complexity - O(max(arr))
class Solution(object):
    def deleteAndEarn(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0
        n = max(nums)
        #print(n)
        tmp = [0]*(n+1)
        for i in nums:
            tmp[i] = tmp[i] + i
        #print(tmp)
        dp = []
        for i in range(n+1):
            dp.append([0,0])
        dp[0][1] = tmp[1]
        for i in range(n+1):
            dp[i][0] = max(dp[i-1][0],dp[i-1][1])
            dp[i][1] = tmp[i] + dp[i-1][0]
        return max(dp[-1][0],dp[-1][1])
            